From 45dd717ed449ae9cc38d126ca8df9ac61541bccf Mon Sep 17 00:00:00 2001
From: Jimmy Xiang <jxiang@apache.org>
Date: Sat, 11 Jun 2016 09:02:45 -0700
Subject: [PATCH 0806/1164] CDH-46822 HIVE-13960: Session may timeout before
 idle timeout time for synchronous operations
 (Zhihai Xu via Jimmy Xiang)

Change-Id: I3d9fd8849f9378b5df54747250ad0106cc256fb7
---
 .../hive/service/cli/session/HiveSessionImpl.java  |   10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java b/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
index 3126eb3..db49521 100644
--- a/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
+++ b/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
@@ -109,6 +109,7 @@
 
   private volatile long lastAccessTime;
   private volatile long lastIdleTime;
+  private volatile int pendingCount = 0;
 
   public HiveSessionImpl(TProtocolVersion protocol, String username, String password,
       HiveConf serverhiveConf, String ipAddress) {
@@ -302,6 +303,8 @@ protected synchronized void acquire(boolean userAccess) {
       lastAccessTime = System.currentTimeMillis();
     }
     Hive.set(sessionHive);
+    pendingCount++;
+    lastIdleTime = 0;
   }
 
   /**
@@ -321,10 +324,11 @@ protected synchronized void release(boolean userAccess) {
     if (userAccess) {
       lastAccessTime = System.currentTimeMillis();
     }
-    if (opHandleSet.isEmpty()) {
+    pendingCount--;
+    // lastIdleTime is only set by the last one
+    // who calls release with empty opHandleSet.
+    if (pendingCount == 0 && opHandleSet.isEmpty()) {
       lastIdleTime = System.currentTimeMillis();
-    } else {
-      lastIdleTime = 0;
     }
   }
 
-- 
1.7.9.5

